tools/ClassGen/NodeBuilder.cpp

```diff
@@ -16,6 +16,169 @@
 
 #include "NodeBuilder.h"
 
+EnumBuilder::EnumBuilder(std::ofstream &C, const std::string &name,
+                         bool isBackendSpecific)
+    : fullyQualifiedName_(name), cStream(C),
+      isBackendSpecific_(isBackendSpecific) {
+  // Parse out and store namespaces if any have been provided by repeatedly
+  // searching for tokens delimited by the scope resolution operator "::"".
+  const std::string kScopeOperator("::");
+
+  size_t namespaceStart = 0;
+  size_t namespaceEnd = name.find(kScopeOperator, namespaceStart);
+
+  while (namespaceEnd != std::string::npos) {
+    // Extract and store the current outermost namespace.
+    const size_t namespaceLength = namespaceEnd - namespaceStart;
+    namespaces_.emplace_back(name.substr(namespaceStart, namespaceLength));
+
+    // Look for the next namespace.
+    namespaceStart = namespaceEnd + kScopeOperator.size();
+    namespaceEnd = name.find(kScopeOperator, namespaceStart);
+  }
+
+  // Once the scope resolution operator is no longer found, that which was
+  // thought to be the beginning of the next namespace must be the beginning
+  // of the enum class name instead.
+  const size_t enumNameStart = namespaceStart;
+  name_ = name.substr(enumNameStart);
+
+  // Sanity check that there are no scope resolution operators in the parsed
+  // enum name.
+  assert(name_.find(kScopeOperator) == std::string::npos);
+}
+
+EnumBuilder &EnumBuilder::addEnumCase(const std::string &name,
+                                      const std::string &doc) {
+  enumCases_.emplace_back(std::make_tuple(name, unusedValue_++, doc));
+  return *this;
+}
+
+EnumBuilder &EnumBuilder::addEnumCaseWithValue(const std::string &name,
+                                               const int value,
+                                               const std::string &doc) {
+  enumCases_.emplace_back(std::make_tuple(name, value, doc));
+  // Maintain the invariant that unusedValue_ is always a value that is NOT
+  // used by the enum.
+  unusedValue_ = std::max(unusedValue_, value + 1);
```

Comment:
This may not matter much but this doesn't prevent multiple cases with the same value if someone makes n cases with `addEnumCase()` then calls `addEnumCaseWithValue()` with value < n expecting that this will automatically avoid conflicts. Is it worth it to keep a set of values and assert that new cases aren't in that?

Reply:
Even just `assert(value >= unusedValue_)` would work then just make `unusedValue_ = value + 1`

Reply:
I sort of fixed this by scanning for and fixing duplicate values in `addEnumCaseWithValue` if a user-defined value conflicts with a value assigned by the tool. It's a linear time method instead of a constant time method now, though.